metrics_sdk/lib/opentelemetry/sdk/metrics/state/asynchronous_metric_stream.rb

```diff
@@ -48,28 +50,53 @@ def collect(start_time, end_time)
           def invoke_callback(timeout, attributes)
             if @registered_views.empty?
               @mutex.synchronize do
-                Timeout.timeout(timeout || 30) do
-                  @callback.each do |cb|
-                    value = cb.call
-                    @default_aggregation.update(value, attributes, @data_points)
-                  end
+                @callback.each do |cb|
+                  value = safe_guard_callback(cb, timeout: timeout)
+                  @default_aggregation.update(value, attributes, @data_points) if value.is_a?(Numeric)
                 end
               end
             else
               @registered_views.each do |view|
                 @mutex.synchronize do
-                  Timeout.timeout(timeout || 30) do
-                    @callback.each do |cb|
-                      value = cb.call
-                      merged_attributes = attributes || {}
-                      merged_attributes.merge!(view.attribute_keys)
-                      view.aggregation.update(value, merged_attributes, @data_points) if view.valid_aggregation?
-                    end
+                  @callback.each do |cb|
+                    value = safe_guard_callback(cb, timeout: timeout)
+                    next unless value.is_a?(Numeric) # ignore if value is not valid number
+
+                    merged_attributes = attributes || {}
+                    merged_attributes.merge!(view.attribute_keys)
+                    view.aggregation.update(value, merged_attributes, @data_points) if view.valid_aggregation?
                   end
                 end
               end
             end
           end
+
+          def now_in_nano
+            (Time.now.to_r * 1_000_000_000).to_i
+          end
```

Comment:
There's a helper for this now in the common gem: https://github.com/open-telemetry/opentelemetry-ruby/blob/3d8243821de54862052d40e5a539a503526bb119/common/lib/opentelemetry/common/utilities.rb#L45-L51